instrumentation/net/http/otelhttpconv/DESIGN.md

```diff
@@ -0,0 +1,283 @@
+	BytesRead() int64
+
+	// Error contains the last error that occurred while reading
+	Error() error
+}
+
+// ResponseWrapper provides a layer on top of `*http.Response` that tracks
+// additional information such as bytes written etc.
+type ResponseWrapper struct {
+	http.ResponseWriter
+
+	// OnWrite is a callback that tracks the amount of data written to the response
+	OnWrite func(n int64)
+
+	// Duration contains the duration of the HTTP request
+	Duration() time.Duration
+
+	// Duration contains the status code returned by the HTTP request
+	StatusCode() int
+
+	// BytesWritten contains the amount of bytes written to the response's body
+	BytesWritten() int64
+
+	// Error contains the last error that occurred while writing
+	Error() error
+}
+```
+
+Note: to keep this document concise, the full implementation is not provided
+here.
+
+### Interfaces
+
+We will provide two public interface that allow interacting with the
+instrumentation package.
+One for client. and one for servers. Implementations can use one, the other or both.
+
+#### The client
+
+```golang
+// Client provides an interface for HTTP client instrumentations to set the
+// proper semantic convention attributes and metrics into their data.
+type Client interface {
+	// RecordError records an error returned by the HTTP request.
+	RecordError(error)
+
+	// Record records the telemetry (spans and metrics) from the provided HTTP request.
+	//
+	// This method does not create a new span. It retrieves the current one from
+	// the context.
+	// It remains the instrumentation's responsibility to start and end spans.
+	Record(ctx context.Context, req *http.Request, w BodyWrapper, cfg ...ClientRecordOption)
+}
+
+// RecordOption applies options to the Record method
+type ClientRecordOption interface{}
+```
+
+#### The Server
+
+```golang
+// Server provides an interface for HTTP server instrumentations to set the
+// proper semantic convention attributes and metrics into their data.
+type Server interface {
+	// Record records the telemetry (spans and metrics) from the provided HTTP
+	// request and response.
+	//
+	// This method does not create a new span. It retrieves the current one from
+	// the context.
+	// It remains the instrumentation's responsibility to start and end spans.
+	Record(ctx context.Context, req *http.Request, w ResponseWrapper, cfg ...ServerRecordOption)
+}
+
+// ServerRecordSpanOption applies options to the RecordSpan method
+type ServerRecordOption interface{}
+```
+
+The `ClientRecordOption` and `ServerRecordOption` functional options allows
+passing optional parameters to be used within the `Record` method, such as the
+HTTP route.
+
+When the data those options provide is not specified, the related span
+attributes will not be set.
+
+### Compatibility
+
+These interfaces are kept relatively simple on purpose, so they can be used as
+foundations if folks wish to extend them with their own actual implementations.
```

Comment:
Do we really want to support custom semconv implementations?
I think it is an overkill (and foodgun). Should we just focus on OTel SemConv?
People not happy with OTel SemConv can create their own instrumentation libraries. 
I propose that `otelhttpconv` has "concrete" implementations (functions, structs).
If we want to provide customization, I would rather add it on instrumentation library layer as this is the API that the user is interacting with.

Reply:
I think being able to support custom semantic conventions is just a side effect of the modularity. The real interesting thing here would be to be able to support unstable semantic conventions, with no breaking changes on the code that supports stable ones.

Reply:
I find it hard to evaluate without code and examples.